cterasdk/core/setup.py

```diff
@@ -0,0 +1,232 @@
+            logging.getLogger().info('Initializing a Replication Database Server. %s', {'host': ipaddr, 'replicate_from': replicate_from})
+            logging.getLogger().debug('Retrieving database replication candidates.')
+            replication_candidates = {re.search('([^/]+$)', k).group(0): k for k in self.get_replication_candidates()}
+            if replication_candidates:
+                server = replication_candidates.get(replicate_from)
+                if server:
+                    logging.getLogger().debug('Found server in replication candidates. %s', {'server': replicate_from})
+                    params = Setup._init_replication_param(server)
+                    self._init_role(params)
+                else:
+                    logging.getLogger().error('Could not find database replication target. %s', {
+                        'target': replicate_from,
+                        'options': replication_candidates
+                    })
+                    raise CTERAException('Could not find database replication target.',
+                                         None, target=replicate_from, options=replication_candidates)
+            else:
+                logging.getLogger().error('Could not find database replication candidates.')
+        self._portal.startup.wait()
+
+    def _init_role(self, params):
+        response = self._portal.execute('/%s/public/servers' % (self._portal.context), 'setReplication', params, use_file_url=True)
+        status = SetupWizardStatusMonitor(self._portal).wait(SetupWizardStage.Replication)
+        self.stage = status.wizard
+        return response
+
+    def get_replication_candidates(self):
+        return self._portal.execute('/%s/public/servers' % (self._portal.context), 'getReplicaitonCandidates', None, use_file_url=True)
+
+    @staticmethod
+    def _init_replication_param(replicate_from=None):
+        params = Object()
+        params._classname = 'SetReplicationParam'  # pylint: disable=protected-access
+        if replicate_from:
+            params.enabledReplicationParam = Object()
+            params.enabledReplicationParam._classname = 'EnabledReplicationParam'  # pylint: disable=protected-access
+            params.enabledReplicationParam.replicationOf = replicate_from
+            params.enabledReplicationParam.restartDB = True
+        return params
+
+    @staticmethod
+    def _init_server_params(mode):
+        params = Object()
+        params._classname = 'InitServerParams'  # pylint: disable=protected-access
+        params.serverMode = mode
+        if mode == ServerMode.Slave:
+            params.slaveSettings = Object()
+            params.slaveSettings._classname = 'SlaveServerSettings'  # pylint: disable=protected-access
+        return params
+
+    def get_setup_status(self):
+        return self._portal.get('/%s/setup/status' % (self._portal.context), use_file_url=True)
+
+    @staticmethod
+    def default_settings():
```

Comment:
Shouldn't the user be able to set all these values?

Reply:
I never liked that we forced users to supply the e-mail settings on initial setup. I prefer this will be handled as a separate module.

Reply:
Leave it be for now